### DISCUSSION TITLE
Bucket notifications to kafka - persistent event store slowness

### DISCUSSION THREAD

**Discussion by vvvlad** [minio/minio#14636] [Q&A] [answered]
Based on this:https://docs.min.io/docs/minio-bucket-notification-guide.html#apache-kafka

When the configuration is similar to this:

`mc admin config set minil notify_kafka:1 tls_skip_verify="off" queue_dir="" queue_limit="100" sasl="off" sasl_password="" sasl_username="" tls_client_auth="0" tls="off" client_tls_cert="" client_tls_key="" brokers="broker:9092" topic="minio" version=""`

the events reach kafka as expected, very fast.

When `queue_dir` value is set, the events seem to be sent very slowly (1 event per few seconds). I do see the json files created in the /home/events directory and no event is lost. 

`mc admin config set minil notify_kafka:1 tls_skip_verify="off" queue_dir="/home/events" queue_limit="100" sasl="off" sasl_password="" sasl_username="" tls_client_auth="0" tls="off" client_tls_cert="" client_tls_key="" brokers="broker:9092" topic="minio" version=""`

Is this an expected behavior or is there something that can be done?

Thanks

**Comment by harshavardhana** [comment, marked as answer]
That means your queue_dir is slow

**Reply by vvvlad** [reply]
Thanks for your reply, but can you please be more specific?
I have kafka docker (confluentinc) running on a VM and the path is local VM storage. 
This means there is no network latency. The storage is very large and fast. 
How would you explain this with slow queue_dir?

**Reply by harshavardhana** [reply]
we write to queue_dir as the events come in and dequeue them in order so what ever your drive is presented is slow for these writes and reading.

**Reply by vvvlad** [reply]
We took some time to explore this and now have some findings I would like to share:
1. We checked various storage options, very fast ones, and that had no effect on the above issue. 
2. The issue seem to be related to the network latency. 

With queue_dir enabled, if kafka instance is located in same region it behaves as expected, but when kafka instance is in a different region (higher latency) than we see the issue.

The issue is not some mild slowness, but a real slowness where messages are reaching kafka in rate of 1-2 messages per second.

Would really appreciate some help with understanding if this is an expected behaviour in this case and if there is a workaround to allow global scaling of this feature.

Thanks again

**Reply by harshavardhana** [reply]
If we cannot push events faster, I don't know what we can do here. Deploy Kafka closer to your deployment.

**Reply by vvvlad** [reply]
when queue_dir is not configured, events are pushed very fast, event for remote kafka.
Without understanding this implementation internals, we could guess that maybe in case of queue_dir implementation instead of opening a single connection, a connection is opened for each event, which can lead to this extra slowness. Or maybe it is a single connection, but minio is waiting for confirmation for each event.

**Reply by harshavardhana** [reply]
Of course we are waiting to ensure that we provide ordered delivery of events.

**Reply by vvvlad** [reply]
yes, but you do that when queue_dir is not set as well, right? so this is probably not the reason, unless the connection is treated differently.

**Reply by harshavardhana** [reply]
> yes, but you do that when queue_dir is not set as well, right? so this is probably not the reason, unless the connection is treated differently.

We do not do that when queue_dir is not set.
